File: packages/ui-swiftui/Sources/KinetixUI/Select.swift
//
// Select.swift — KinetixSelect.
//
// Mirrors packages/ui/src/components/select.tsx. The Radix trigger +
// portalled listbox collapses onto a native `Menu` with a styled trigger
// label (border / chevron matching KinetixInput) — the same "reuse the
// platform machinery" call the Compose port made building on Material3's
// `DropdownMenu`. `min-h-[44px]` is Figma-literal. Options are a value +
// label list; selection is a `Value?` binding.
//

import SwiftUI

public struct KinetixSelectOption<Value: Hashable>: Identifiable {
    public let id: Value
    public let label: String
    public var value: Value { id }

    public init(value: Value, label: String) {
        self.id = value
        self.label = label
    }
}

public struct KinetixSelect<Value: Hashable>: View {
    @Environment(\.kinetixColors) private var colors
    @Environment(\.isEnabled) private var isEnabled

    @Binding private var selection: Value?
    private let options: [KinetixSelectOption<Value>]
    private let placeholder: String
    private let isError: Bool

    public init(
        selection: Binding<Value?>,
        options: [KinetixSelectOption<Value>],
        placeholder: String = "Select…",
        isError: Bool = false
    ) {
        self._selection = selection
        self.options = options
        self.placeholder = placeholder
        self.isError = isError
    }

    private var currentLabel: String {
        options.first { $0.value == selection }?.label ?? placeholder
    }

    public var body: some View {
        Menu {
            ForEach(options) { opt in
                Button {
                    selection = opt.value
                } label: {
                    if selection == opt.value {
                        Label(opt.label, systemImage: "checkmark")
                    } else {
                        Text(opt.label)
                    }
                }
            }
        } label: {
            HStack {
                Text(currentLabel)
                    .font(.system(size: 14))
                    .foregroundStyle(selection == nil ? colors.mutedForeground : colors.foreground)
                    .lineLimit(1)
                Spacer(minLength: 8)
                Image(systemName: "chevron.down")
                    .font(.system(size: 14))
                    .foregroundStyle(colors.foreground.opacity(0.6))
            }
            .padding(.horizontal, 12) // spacing/3
            .padding(.vertical, 12)
            .frame(minHeight: 44, alignment: .leading)
            .frame(maxWidth: .infinity)
            .background(colors.background, in: RoundedRectangle(cornerRadius: 4, style: .continuous)) // radius/sm
            .overlay {
                RoundedRectangle(cornerRadius: 4, style: .continuous)
                    .strokeBorder(isError ? colors.destructive : colors.input, lineWidth: 1)
            }
            .contentShape(Rectangle())
        }
        .opacity(isEnabled ? 1 : 0.5)
    }
}

File: packages/ui-flutter/lib/src/select.dart
import 'package:flutter/material.dart';

import 'app_text.dart';
import 'theme.dart';

class KinetixSelectOption<T> {
  const KinetixSelectOption(this.value, this.label);

  final T value;
  final String label;
}

/// Mirrors `packages/ui/src/components/select.tsx`. The Radix trigger +
/// portalled listbox collapses onto a `MenuAnchor` with a styled trigger
/// (border / chevron like KinetixInput) — the "reuse the platform
/// machinery" call. `min-h-[44px]` is Figma-literal.
class KinetixSelect<T> extends StatelessWidget {
  const KinetixSelect({
    super.key,
    required this.value,
    required this.options,
    required this.onChanged,
    this.placeholder = 'Select…',
    this.isError = false,
  });

  final T? value;
  final List<KinetixSelectOption<T>> options;
  final ValueChanged<T> onChanged;
  final String placeholder;
  final bool isError;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);

    var current = placeholder;
    for (final o in options) {
      if (o.value == value) current = o.label;
    }

    return MenuAnchor(
      style: MenuStyle(
        backgroundColor: WidgetStatePropertyAll(c.popover),
        surfaceTintColor: const WidgetStatePropertyAll(Color(0x00000000)),
        shape: WidgetStatePropertyAll(
          RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
            side: BorderSide(color: c.border),
          ),
        ),
      ),
      menuChildren: [
        for (final o in options)
          MenuItemButton(
            onPressed: () => onChanged(o.value),
            leadingIcon: o.value == value
                ? Icon(Icons.check, size: 16, color: c.foreground)
                : const SizedBox(width: 16),
            child: Text(o.label, style: AppText.bodyMd.copyWith(color: c.foreground)),
          ),
      ],
      builder: (context, controller, _) => GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: () => controller.isOpen ? controller.close() : controller.open(),
        child: Container(
          constraints: const BoxConstraints(minHeight: 44),
          padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 12),
          decoration: BoxDecoration(
            color: c.background,
            borderRadius: BorderRadius.circular(4), // radius/sm
            border: Border.all(color: isError ? c.destructive : c.input, width: 1),
          ),
          child: Row(
            children: [
              Expanded(
                child: Text(
                  current,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: AppText.bodyMd.copyWith(color: value == null ? c.mutedForeground : c.foreground),
                ),
              ),
              Icon(Icons.keyboard_arrow_down, size: 18, color: c.mutedForeground),
            ],
          ),
        ),
      ),
    );
  }
}
